### PR TITLE
feat: quant_llm support

### PR DESCRIPTION
**Pull request by AlpinDale**
Based on #617 by @intervitens

**TL;DR** load an FP16 model in bit sizes between 2 and 7: `aphrodite run <your model> -q fp4` for example.

This PR adds support for online (on-the-fly, runtime) quantization of FP16 models to FP2, FP3, FP4, FP5, FP6, and FP7. Unlike DeepSpeedFP, this method is actually fast and has a wider range of hardware support.

Does not need to load the model in FP16 first before quantization, unlike `-q fp8` and DeepSpeedFP.

For more fine-grained control over the data type, you can specify the exponent to use:

`-q quant_llm -quant-llm-fp-bits {2,3,4,5,6,7} --quant-llm-exp-bits {1,2,3,4,5}`

The heuristic used for determining the mantissa value is `weight_bits - exp_bits - 1`. Make sure your provided values are valid based on that, and should not result in a negative number. See [here](https://gist.github.com/AlpinDale/17babab5be16f522d4d3b134e171001a) for the list of all valid combinations.

Kernels have been ported and adapted directly within Aphrodite.

### Accuracy
**FP4**
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.6528|±  |0.0131|
|     |       |strict-match    |     5|exact_match|↑  |0.6429|±  |0.0132|

**FP5**
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.7680|±  |0.0116|
|     |       |strict-match    |     5|exact_match|↑  |0.7506|±  |0.0119|

**FP6**
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.7854|±  |0.0113|
|     |       |strict-match    |     5|exact_match|↑  |0.7688|±  |0.0116|

**FP7**
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.7718|±  |0.0116|
|     |       |strict-match    |     5|exact_match|↑  |0.7551|±  |0.0118|

***

**For comparison**:

**BF16**
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.7794|±  |0.0114|
|     |       |strict-match    |     5|exact_match|↑  |0.7574|±  |0.0118|

**FP8 (Marlin)**
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.7733|±  |0.0115|
|     |       |strict-match    |     5|exact_match|↑  |0.7536|±  |0.0119|

**NF4 (bitsandbytes)**
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.7309|±  |0.0122|
|     |       |strict-match    |     5|exact_match|↑  |0.7263|±  |0.0123|

**AWQ INT4 (Marlin)**
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.7437|±  |0.0120|
|     |       |strict-match    |     5|exact_match|↑  |0.7346|±  |0.0122|

**AWQ INT4**
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.7400|±  |0.0121|
|     |       |strict-match    |     5|exact_match|↑  |0.7316|±  |0.0122|

**GPTQ (Marlin)** (what went wrong here?)
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.3935|±  |0.0135|
|     |       |strict-match    |     5|exact_match|↑  |0.3927|±  |0.0135|

**GPTQ**
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.7240|±  |0.0123|
|     |       |strict-match    |     5|exact_match|↑  |0.7202|±  |0.0124|

### Throughput
**NVIDIA A40**

```console
quantization=fp2
python tests/benchmarks/engine/throughput.py --model NousResearch/Meta-Llama-3.1-8B-Instruct -q $quantization --max-model-len 2048 --input-len 512 --output-len 52
```

| Data Type | Request/s | Tokens/s | Weight Mem Usage (GiB) |
| ---- | ----  | ---- | ---- |
| FP2_E1M0 | 10.88 | 6139.12 | 3.62 |
| FP3_E2M0 | 10.09 | 5693.54 | 4.46 |
| FP4_E2M1 | 10.11 | 5700.73 | 5.24 |
| FP5_E2M2 | 9.49 | 5354.48 | 6.11 |
| FP6_E2M3 | 9.46 | 5337.18 | 6.87 |
| FP7_E3M3 | 9.12 | 5142.04 | 7.74 |
| FP8_E4M3 (Marlin) | 6.03 | 3403.61 | 8.49 |
| FP16_E5M10 (Unquantized) | 10.98 | 6190.34 | 14.99 |
| FP16_E8M7 (BF16) | 10.98 | 6190.34 | 14.99 |

FP2 and FP3 not recommended.

### CODE REVIEW

**Comment by ZanePoe** [comment]
Does this quantitative reasoning method support the GPU of the Turing architecture?like 2080ti
